panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,513 @@
+  const auto cc_gain2 = info_.hardware_parameters.at("cc_gain2");
+  const auto cc_t0 = info_.hardware_parameters.at("cc_t0");
+  const auto cc_t1 = info_.hardware_parameters.at("cc_t1");
+  const auto cc_t2 = info_.hardware_parameters.at("cc_t2");
+  const auto cc_t3 = info_.hardware_parameters.at("cc_t3");
+  const auto cc_t4 = info_.hardware_parameters.at("cc_t4");
+  const auto cc_t5 = info_.hardware_parameters.at("cc_t5");
+
+  params_.cc_mag_field = hardware_interface::stod(cc_mag_field);
+  params_.cc_offset0 = hardware_interface::stod(cc_offset0);
+  params_.cc_offset1 = hardware_interface::stod(cc_offset1);
+  params_.cc_offset2 = hardware_interface::stod(cc_offset2);
+  params_.cc_gain0 = hardware_interface::stod(cc_gain0);
+  params_.cc_gain1 = hardware_interface::stod(cc_gain1);
+  params_.cc_gain2 = hardware_interface::stod(cc_gain2);
+  params_.cc_t0 = hardware_interface::stod(cc_t0);
+  params_.cc_t1 = hardware_interface::stod(cc_t1);
+  params_.cc_t2 = hardware_interface::stod(cc_t2);
+  params_.cc_t3 = hardware_interface::stod(cc_t3);
+  params_.cc_t4 = hardware_interface::stod(cc_t4);
+  params_.cc_t5 = hardware_interface::stod(cc_t5);
+}
+
+void PantherImuSensor::ReadMadgwickFilterParams()
+{
+  const auto gain = info_.hardware_parameters.at("gain");
+  const auto zeta = info_.hardware_parameters.at("zeta");
+  const auto mag_bias_x = info_.hardware_parameters.at("mag_bias_x");
+  const auto mag_bias_y = info_.hardware_parameters.at("mag_bias_y");
+  const auto mag_bias_z = info_.hardware_parameters.at("mag_bias_z");
+  const auto use_mag = info_.hardware_parameters.at("use_mag");
+  const auto stateless = info_.hardware_parameters.at("stateless");
+  const auto remove_gravity_vector = info_.hardware_parameters.at("remove_gravity_vector");
+
+  CheckMadgwickFilterWorldFrameParam();
+
+  params_.gain = hardware_interface::stod(gain);
+  params_.zeta = hardware_interface::stod(zeta);
+  params_.mag_bias_x = hardware_interface::stod(mag_bias_x);
+  params_.mag_bias_y = hardware_interface::stod(mag_bias_y);
+  params_.mag_bias_z = hardware_interface::stod(mag_bias_z);
+  params_.use_mag = hardware_interface::parse_bool(use_mag);
+  params_.stateless = hardware_interface::parse_bool(stateless);
+  params_.remove_gravity_vector = hardware_interface::parse_bool(remove_gravity_vector);
+}
+
+void PantherImuSensor::CheckMadgwickFilterWorldFrameParam()
+{
+  const auto world_frame = info_.hardware_parameters.at("world_frame");
+
+  auto it = kImuWorldFramesMap.find(world_frame);
+  if (it == kImuWorldFramesMap.end()) {
+    throw std::runtime_error(
+      "The parameter world_frame was set to invalid value. "
+      "Valid values are 'enu', 'ned' and 'nwu'. Setting to 'enu'.");
+  }
+  world_frame_ = it->second;
+}
+
+void PantherImuSensor::SetInitialValues()
+{
+  imu_sensor_state_.resize(
+    info_.sensors[0].state_interfaces.size(), std::numeric_limits<double>::quiet_NaN());
+}
+
+void PantherImuSensor::Calibrate()
```

Comment:
Should we block the read functionality when the calibration callback is executed, or is there something else we should do?

Reply:
We decided to calibrate the sensor before reading the sensor!